pkg/nsx/services/firewall.go

```diff
@@ -1322,58 +1361,135 @@ func (service *SecurityPolicyService) CreateOrUpdateSecurityPolicy(obj *v1alpha1
 	for _, rule := range indexResults {
 		existingRules = append(existingRules, rule.(model.Rule))
 	}
-	if service.securityPolicyEqual(&existingSecurityPolicy, nsxSecurityPolicy) && service.rulesEqual(existingRules, nsxSecurityPolicy.Rules) {
-		log.Info("security policy not changed, skipping", "nsxSecurityPolicy.Id", nsxSecurityPolicy.Id)
+	// Caution! createOrUpdate can't delete the legacy rules.
+	// So we have to delete the legacy rules firstly by rulesClient.
+	spEqual := service.securityPolicyEqual(&existingSecurityPolicy, nsxSecurityPolicy)
+	ruleEqual, legacyRules := service.rulesEqual(existingRules, nsxSecurityPolicy.Rules)
+	if spEqual && ruleEqual {
+		log.Info("security policy and rules not changed, skip", "nsxSecurityPolicy.Id", nsxSecurityPolicy.Id)
 	} else {
-		err = policiesClient.Patch(service.getDomain(), *nsxSecurityPolicy.Id, *nsxSecurityPolicy)
+		err := service.updateOrDeleteRules(nsxSecurityPolicy, legacyRules)
 		if err != nil {
-			log.Error(err, "failed to patch security policy", "nsxSecurityPolicy", nsxSecurityPolicy)
 			return err
 		}
-		service.SecurityPolicyStore.Add(*nsxSecurityPolicy)
-		for _, rule := range nsxSecurityPolicy.Rules {
-			service.RuleStore.Add(rule)
+		err = service.createOrUpdateSecurityPolicy(nsxSecurityPolicy)
+		if err != nil {
+			return err
 		}
 		log.Info("successfully created or updated nsxSecurityPolicy", "nsxSecurityPolicy", nsxSecurityPolicy)
 	}
+
+	// The reason why delete legacy groups at last is that some rules may have reference to the legacy groups.
```

Comment:
Why not put line 1337-1347 in the end here?

Reply:
The added groups must appear before the added rules in case some rules' groups referred to them, otherwise, it would report an error that the groups don't exist.